konnectivity-client/pkg/client/client.go

```diff
@@ -130,9 +132,9 @@ func (t *grpcTunnel) serve() {
 				t.connsLock.Lock()
 				delete(t.conns, resp.ConnectID)
 				t.connsLock.Unlock()
-			} else {
-				klog.Warningf("connection id %d not recognized", resp.ConnectID)
+				return
```

Comment:
Using return is cleaner than break because we would need to break out of both the `switch` statement and the `for` loop.

Reply:
The grpcTunnel is meant to be a persistent "tunnel". Clients will only create the grpcTunnel once, and every time they need to communicate to a remote host, then call the grpcTunnel.Dial(). The tunnel supports multiple communication streams because these streams are uniquely identified by the "connectionID".

In this sense, when one stream is closed, we shouldn't close the tunnel, because there could be other ongoing streams multiplexed on this tunnel.

Kubernetes doesn't use the grpcTunnel this way. When the apiserver needs a network connection to the remote, it always creates a new grpc connection to the proxy server, sends a dial request over the tunnel and gets a connection back, see [1](https://github.com/kubernetes/kubernetes/blob/695cd49c40fb174728245814fd2db25f9b801d42/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go#L228-L233).

I think another to fix the too many open UDS connection is to fix Kubernetes. The apiserver should maintain and reuse a single grpcTunnel. This might bring some performance improvements as well. This would be a big change.

Alternatively, as a quick fix, we can create a "singleUseGRPCTunnel" object, and make sure that clients can only use this tunnel to dial once. And docuement that when the connection over this tunnel is closed, the tunnel will also be closed.

Reply:
That makes sense. I created https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/101 and will follow up. I like the "singleUseGRPCTunnel" idea, but we will probably need more testing for failure conditions (connection broken, health checks, etc). In the short run, is this PR needed to avoid creating too many streams/tunnels?

Reply:
This PR is not mergeable as it is.

> but we will probably need more testing for failure conditions (connection broken, health checks, etc)?

Why do we need more testing? In case we are not on the same page, in my mind, the `singleUseGRPCTunnel` will be very similar to the `grpcTunnel`, but instead of managing a pool of pending dials and a pool of connections, it only manages one pending dial and one connection. And it stops when it receives the CLOSE_RSP.

Reply:
Created the `singleUseGRPCTunnel` object and renamed the tunnel creation function to `CreateSingleUseGRPCTunnel`. I will do some more investigation and testing with https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/101